rofl-proxy/src/http/sni.rs

```
@@ -0,0 +1,181 @@
+
+    // Skip session ID.
+    let length = buf.read_u8()?;
+    if length > 32 {
+        return Err(io::Error::other("corrupted session ID"));
+    }
+    buf.seek_relative(length as i64)?;
+    // Skip cipher suite.
+    let length = buf.read_u16::<BigEndian>()?;
+    if length < 2 || length % 2 != 0 {
+        return Err(io::Error::other("corrupted cipher suite"));
+    }
+    buf.seek_relative(length as i64)?;
+    // Skip compression method.
+    let length = buf.read_u8()?;
+    if length < 1 {
+        return Err(io::Error::other("corrupted compression method"));
+    }
+    buf.seek_relative(length as i64)?;
+
+    // Extensions.
+    let _extensions_length = buf.read_u16::<BigEndian>()?;
+    loop {
+        // Extension header:
+        //   type:   u16
+        //   length: u16
+        let extension_type = buf.read_u16::<BigEndian>()?;
+        let extension_length = buf.read_u16::<BigEndian>()?;
+        if extension_type != TLS_EXTENSION_TYPE_SNI {
+            // Skip over non-SNI extensions.
+            buf.seek_relative(extension_length as i64)?;
+            continue;
+        }
+
+        // SNI extension header:
+        //   server name list length: u16
+        let _list_length = buf.read_u16::<BigEndian>()?;
+        loop {
+            let name_type = buf.read_u8()?;
+            let name_length = buf.read_u16::<BigEndian>()?;
+            if name_type != TLS_SNI_TYPE_HOSTNAME {
+                // Skip over non-hostname SNI.
+                buf.seek_relative(name_length as i64)?;
+                continue;
+            }
+
+            let name_length = name_length as usize;
+            if buf.get_ref().len() - (buf.position() as usize) < name_length {
+                return Err(io::Error::other("corrupted SNI extension: bad name length"));
+            }
+            let buf = &buf.get_ref()[buf.position() as usize..];
+            let name = String::from_utf8_lossy(&buf[..name_length]);
```

Should this instead use `from_utf8` and fail in case of invalid utf-8?